unit tests for readiness controller

irbekrm opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
4. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What this PR does / why we need it**:

1. Adds a unit test for readiness controller's ProcessItem method
2. Adds a unit test for readiness controller's policy chain
3. Refactors the policies a little to make them easier to understand <- this is subjective
4. Moves the calculation of certificate's renewal time into  [certificates package](https://github.com/jetstack/cert-manager/tree/master/pkg/controller/certificates) to make it easier to use and test

I was going to also add an integration test for the readiness controller, however, this PR is already size/XL so I will make a separate PR for that.

At some point it might make sense to move [policies](https://github.com/jetstack/cert-manager/tree/master/pkg/controller/certificates/trigger/policies) out of the trigger package as they are used in readiness controller too.


fixes #3666 


```release-note
NONE
```
/kind cleanup

irbekrm commented:
I'm updating state of this PR to WIP as I've realized that it'd be cleaner and easier to merge/cherrypick #3665  first.

JoshVanL commented on pkg/controller/certificates/trigger/policies/policies.go:
```diff
@@ -44,13 +45,13 @@ type Func func(Input) (reason, message string, reissue bool)
-		reason, message, reissue := policyFunc(input)
-		if reissue {
-			return reason, message, reissue
+		reason, message, violationFound := policyFunc(input)
```
The bool variable rename is much better :smile:

JoshVanL commented on pkg/controller/certificates/trigger/policies/policies.go:
```diff
@@ -189,27 +190,29 @@ func CurrentCertificateNearingExpiry(c clock.Clock) Func {
-	}
+func CurrentCertificateHasExpired(c clock.Clock) Func {
+	return func(input Input) (string, string, bool) {
+		certData := input.Secret.Data[corev1.TLSCertKey]
```
We should check the bool 'ok' var here too, and exit early if it is false.

irbekrm replied:
I was undecisive about this as theoretically that should have been caught by other policy functions earlier in the chain. But yes, safer to check.

JoshVanL commented on pkg/controller/certificates/trigger/policies/policies_test.go:
```diff
@@ -27,7 +26,7 @@ import (
 	cmapi "github.com/jetstack/cert-manager/pkg/apis/certmanager/v1"
 	cmmeta "github.com/jetstack/cert-manager/pkg/apis/meta/v1"
-	"github.com/jetstack/cert-manager/pkg/util/pki"
+	testutil "github.com/jetstack/cert-manager/pkg/controller/certificates/internal/test"
```
nit: internaltest

JoshVanL commented on pkg/controller/certificates/util.go:
```diff
@@ -262,6 +262,20 @@ func GenerateLocallySignedTemporaryCertificate(crt *cmapi.Certificate, pkData []
+
+// RenewalTimeWrapper returns RenewalTimeFunc implementation
+// TODO: potentially merge RenewBeforeExpiryDuration into this function and rewrite the tests accordingly
+func RenewalTimeWrapper(defaultRenewBeforeExpiryDuration time.Duration) RenewalTimeFunc {
```
I don't really like the wrapper we have here, makes things a bit more confusing then they have to be. Can we not just have these as standard functions? What is t he purpose of `RenewalTimeFunc`?

irbekrm replied:
It is like that because I couldn't find a cleaner way to access [DefaultRenewBefore](https://github.com/jetstack/cert-manager/blob/master/pkg/apis/certmanager/v1/const.go#L32).

Having this as a standard function would mean that that value to be:

a) passed into the controller separately and then passed into the renewal time func when it's called which seems unnecessary passing it around since it's a const anyway
b) accessed directly from the renewal time func which seems like it would make a util function too 'clever'
c) accessed directly from inside controller's ProcessItem function which again seems 'unclean' since it's an external value and we might still want to mock it in tests.

Accessing it from a closure seemed similar to how we already do it in other places, i.e when accessing the clock in [trigger controller's policy func](https://github.com/jetstack/cert-manager/blob/master/pkg/controller/certificates/trigger/trigger_controller.go#L247)

Maybe I'm just overthinking it though 😄

JoshVanL reviewed (changes requested):
Few comments from me.

/assign @irbekrm

JoshVanL commented:
As an aside, not for this PR.. The use of the word policy in this part of the code will makes things quite confusing when we introduce CR policy resources and code. We should think about moving this word over to something like "compliance".

irbekrm commented on pkg/controller/certificates/readiness/readiness_controller_test.go:
```diff
@@ -0,0 +1,461 @@
+				Message: "ready message",
+			},
+			certShouldExist:   true,
+			certShouldUpdate:  true,
```
> Has a ready condition of true and policy passes (should be no update)

This is actually currently not the case- we attempt to update every time. I created an issue for that and was going to fix it after these tests are merged. #3663

irbekrm commented:
Thanks for review @JoshVanL ! I've addressed all the comments apart from the wrapper functions one as I am still not sure whether another approach would be cleaner there, but would be happy to change it if that is the case.

JoshVanL commented:
/lgtm
